docs/src/docs/asciidoc/user/resilience.adoc

```diff
@@ -0,0 +1,189 @@
+* Sleep time between retries (default is 0)
+
+Note that for each retry, the timeout will be enforced. So, far example, 3 retries with a timeout of 5 seconds and a sleep
+of 1 second will mean 20 seconds before returning.
+
+*Note to myself:* Should the retry be on the resilience strategy?
+
+=== In case of failure
+
+If we failed to get an answer, the cache [#Resilience Strategy] will kick in and return whatever it was configured to return.
+
+Resilience strategy is not used for lifecycle events. So in this case you will get a `StoreAccessException`.
+
+=== After a failure
+
+We will now consider a server with no mirror.
+
+The first thing that can happen is a server that has problem answering correctly all the time but that is still there.
+In this case, after 10 timeouts (configurable), the client will consider the server down to prevent waiting on requests.
+It will perform heartbeats in background to determine when the situation goes back to normal and reconnect if needed.
+Meanwhile, it will rely on the resilience strategy to answer.
+
+*Note:* M&M will trigger an alarm if a server is acting erratically.
+
+It is important to notice that the client won't fallback to the resilience strategy when the caching tier answers.
+It also means that the caching tier might not receive updates from the server and become out of sync.
+
+This is independent of the consistency configured and configurable. You can pick the following strategies:
+
+* Rely on tier forever, so even if the server is officially lost
+* Rely on tier on hiccups. The server will keep the caching tier until it declares the server lost for good
+* Don't rely on tier.
+
+When the server is lost, the client won't try to call it until the heartbeat resumes. A nice feature would be to have a
+ledger pointer in the heartbeat. To tell the client is its out of sync when an answer comes.
+
+=== Active and mirror setup
+
+It works the same as with a single server. Except that during a failover, the client will behaves like if the underlying
+server is having hiccups. Or is down if the failover takes too long.
+
+The new active server will notify the client when ready.
+
+=== Reconnection
+
+When a server goes down and back again. On the same URL. The client will silently reconnect to it. The ledger should allow
+to invalidate entries on the caching tier.
+
+== Loader writer
+
+Cache loader writers have a dedicated resilience strategy. So in case of failure, the resilience strategy will operate and
+the result will be picked up by the Cache resilience strategy.
+
+When using a loader writer, we tend to consider the following:
+
+* We expect the cache never return null. Because the loader will load any missing data. But to enforce that
+in case of server failure, it means the client should wait until the backend is back.
+* We expect that everything cached is saved. So in general an exception should be thrown if we fail to do so.
+
+The default resilience strategy is the following.
```

Comment:
I figure we should be able to make the behavior with loader-writers equivalent to the normal behavior of the cache when used in a sensibly coded cache-aside setup.  Is that what you are aiming for here?

Reply:
Tying a cache topology choice to a cache consistency choice seems strange.  Why can't a cache-through/cache-behind user have the same rules as a cache-aside user?

Reply:
They can. Where am I saying they can't? But you are right, the text is highly focused on a single loader writer usage. I need to had some subtlety.

Reply:
Ok. My loader-writer part was actually more complete than what I remember. So I'm not sure what puzzles you. But I think it's my bad writing. There are two resilience strategies:
* CacheResilienceStrategy
* LoaderWriteResilienceStrategy

Here, I am talking about the later.

So, for instance, if a read from the SoR fails, the `LoaderWriteResilienceStrategy` will kick-in and return a result. Which will be received by the `CacheResilienceStrategy` that will return the final result.

My goal is to simplify the resilience strategy by actually having too. And if you are no using a loader-writer. On won't need to know about its resilience strategy. This removes a lot of noise from cache resilience strategy.